src/packages/text_search.gleam

```diff
@@ -46,45 +48,113 @@ pub fn update(
   insert(index, name, description)
 }
 
+pub type LookupOutcome {
+  Packages(packages: List(String))
+  DidYouMean(suggestion: String)
+}
+
 pub fn lookup(
   index: TextSearchIndex,
   phrase: String,
-) -> Result(List(String), Error) {
+) -> Result(LookupOutcome, Error) {
   let phrase = string.lowercase(phrase)
+
   stem_words(phrase)
   |> list.flat_map(expand_search_term)
   |> list.try_map(ethos.get(index.table, _))
-  |> result.map(fn(names) {
-    names
-    |> list.flatten
-    |> list.fold(dict.new(), fn(counters, name) {
-      dict.upsert(counters, name, fn(x) { option.unwrap(x, 0) + 1 })
-    })
+  |> result.replace_error(error.EtsTableError)
+  |> result.try(fn(packages) {
+    case list.flatten(packages) {
+      [_, ..] as packages ->
+        Ok(Packages(packages: rank_results(packages, phrase)))
+
+      // If there's no results matching the given query we try and guess what
+      // the user might have wanted to type.
+      [] -> {
+        use words <- result.try(
+          ethos.keys(index.table)
+          |> result.replace_error(error.EtsTableError),
+        )
+        case did_you_mean(given: phrase, one_of: words) {
+          Ok(suggestion) -> Ok(DidYouMean(suggestion:))
+          Error(_) -> Ok(Packages(packages: []))
+        }
+      }
+    }
+  })
+}
+
+fn did_you_mean(
+  given phrase: String,
+  one_of words: List(String),
+) -> Result(String, Nil) {
+  // We want to limit the maximum edit distance. Otherwise we could end up
+  // suggesting fixes that are not related at all to the original query.
+  let phrase_length = string.length(phrase)
+  let limit = int.max(1, phrase_length / 3)
+
+  list.filter_map(words, fn(word) {
+    let word_length = string.length(word)
+    let minimum_distance = int.absolute_value(word_length - phrase_length)
+
+    // If the minimum distance is greater than the allowed limit then we don't
+    // even waste any time computing the edit distance of the two strings!
+    use <- bool.guard(when: minimum_distance > limit, return: Error(Nil))
+    let distance = edit_distance.levenshtein(phrase, word)
```

Comment:
We said we were gunna use the same algorithm as the compiler, is using the more basic levenshtein algorithm gunna be impactful?

Reply:
The difference between the two is this simpler one won't consider swapping two adjacent letters as a single edit (while the more advanced one will). Take "lsutre" as an example:
- for the simpler Levenshtein implementation that's 2 edits away from "lustre": change `s -> u` and `u -> s`
- for the more complex one (Levenshtein Damerau) it is 1 edit away from "lustre": swap `s <-> u`

It is nice as it can catch simple typos as a single edit rather than two (at the cost of being a bit heavier to compute, especially memory-wise). I don't know how impactful that will be in practice; I went for the simpler one as I had an implementation ready and implementing the other one would require me more time and to understand the other algorithm as well 😆